dom-fetch-client/src/main/scala/org/http4s/dom/FetchClient.scala

```diff
@@ -49,27 +49,29 @@ object FetchClient {
 
           Fetch.fetch(req.uri.renderString, init)
         }
+      }.toResource
+      status <- F.fromEither(Status.fromInt(response.status)).toResource
+      body <- Resource.makeCase(response.body.pure[F]) {
```

Comment:
I think in order for this to work they way you want, you need to have the `response` be in the acquisition of a Resource. If I'm reading this correctly, you want to ensure that if you get a response, that you always register a `_.cancel` on the body. In order to do this you need to acquire the `response` and register the finalizers to call `_.cancel` in the same `Resource`. In your current encoding, for example, if parsing the `status` fails you will never register the finalizers. 

I think you want something like this,

```scala
  def apply[F[_]](implicit F: Async[F]): Client[F] = Client[F] { req: Request[F] =>
    for {
      body <- req.body.chunkAll.filter(_.nonEmpty).compile.last.toResource
      requestInit <- F.catchNonFatal {
          val init = new RequestInit {}

          init.method = req.method.name.asInstanceOf[HttpMethod]
          init.headers = new Headers(toDomHeaders(req.headers))
          body.foreach { body =>
            init.body = arrayBuffer2BufferSource(body.toJSArrayBuffer)
          }
          init
        }.toResource
      response <- Resource.makeCase(F.fromPromise(F.delay(Fetch.fetch(req.uri.renderString, requestInit)))) {
        case (r, Resource.ExitCase.Succeeded) =>
          F.fromPromise(F.delay(r.body.cancel(null))).void
        case (r, Resource.ExitCase.Errored(ex)) =>
          F.fromPromise(F.delay(r.body.cancel(ex.getMessage()))).void
        case (r, Resource.ExitCase.Canceled) =>
          F.fromPromise(F.delay(r.body.cancel(null))).void
      }
      status <- F.fromEither(Status.fromInt(response.status)).toResource
    } yield Response[F](
      status = status,
      headers = fromDomHeaders(response.headers),
      body = readableStreamToStream(response.body)
    )
  }
```

FWIW, in cases like this where we introduce `ExitCase` based finalizers, I personally find for comps more difficult to read than a `flatMap` based approach, but that is inherently subjective.

Reply:
Oops, good catch! You are right, I'll fix this. And I agree, not crazy about the for comprehension here myself, I'll try it the other way.